app/models/manager_refresh/save_collection/helper.rb

```diff
@@ -14,92 +14,92 @@ def save_inventory_object_inventory(ems, inventory_collection)
-      unique_index_keys = inventory_collection.manager_ref_to_cols
+    def save_inventory_collection!(inventory_collection, association)
+      attributes_index        = {}
+      inventory_objects_index = {}
+      inventory_collection.each do |inventory_object|
+        attributes = inventory_object.attributes(inventory_collection)
+        index      = inventory_object.manager_uuid
 
-      association.find_each do |record|
-        index = inventory_collection.object_index_with_keys(unique_index_keys, record)
-        if record_index[index]
-          # We have a duplicate in the DB, destroy it. A find_each method does automatically .order(:id => :asc)
-          # so we always keep the oldest record in the case of duplicates.
-          _log.warn("A duplicate record was detected and destroyed, inventory_collection: '#{inventory_collection}', "\
-                    "record: '#{record}', duplicate_index: '#{index}'")
-          record.destroy
-        else
-          record_index[index] = record
-        end
+        attributes_index[index]        = attributes
+        inventory_objects_index[index] = inventory_object
       end
 
+      unique_index_keys = inventory_collection.manager_ref_to_cols
+      unique_db_indexes = Set.new
+
       inventory_collection_size = inventory_collection.size
+      deleted_counter           = 0
       created_counter           = 0
-      _log.info("*************** PROCESSING #{inventory_collection} of size #{inventory_collection_size} ***************")
+      _log.info("*************** PROCESSING #{inventory_collection} of size #{inventory_collection_size} *************")
       ActiveRecord::Base.transaction do
-        inventory_collection.each do |inventory_object|
-          hash   = inventory_object.attributes(inventory_collection)
-          record = record_index.delete(inventory_object.manager_uuid)
-          if record.nil?
-            next unless inventory_collection.create_allowed?
-            record = inventory_collection.model_class.create!(hash.except(:id))
-            created_counter += 1
+        association.find_each do |record|
+          index = inventory_collection.object_index_with_keys(unique_index_keys, record)
+          if unique_db_indexes.include?(index) # Include on Set is O(1)
+            # We have a duplicate in the DB, destroy it. A find_each method does automatically .order(:id => :asc)
+            # so we always keep the oldest record in the case of duplicates.
+            _log.warn("A duplicate record was detected and destroyed, inventory_collection: '#{inventory_collection}', "\
+                      "record: '#{record}', duplicate_index: '#{index}'")
+            record.destroy
+          else
+            unique_db_indexes << index
+          end
+
+          inventory_object = inventory_objects_index.delete(index)
+          hash             = attributes_index.delete(index)
+
+          if inventory_object.nil?
```

Comment:
this confuses me
I expect if `inventory_object` exists, then to possible delete.

could you put a comment above this check?